---
title: "Tree and tree to map"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Load Data

First we load the required packages and files.

Note that I use a cladogram which I manually constructed with FigTree. The coordinates for the Outgroups are just fake numbers (coordinates for a lake in Turkmenistan) so they align nicely next to our samples.

```{r Load Data, results='hide', message=FALSE}
rm(list = ls())

library(phytools)
library(ggplot2)
library(mapdata)
library(viridis)
library(ggtree)
library(extrafont)
library(cowplot)

Nt_Tree = read.tree("../00_Data/Trees/RAxML_Nucleotide_Cladogram.nwk")

Coordinates = as.matrix(read.csv("../00_Data/Miscellaneous/Coordinates.csv", sep = "\t", header = T, row.names = 1))
```

# Plot phylogeny projected on a map
## Build PhyloToMap object

The xlim and ylim are roughly the border coordinates for Europe. Then we build a color vector based on the lineage (Turkey, viridis, bilineata, Adriatic, outgroup).

```{r Build Phylo To Map object, results='hide', message=FALSE}
obj = phylo.to.map(Nt_Tree, 
                   Coordinates, 
                   xlim = c(-10, 59), 
                   ylim = c(30, 71),
                   fill = T, 
                   col = "moccasin",
                   plot=F, 
                   rotate = T, 
                   direction = "downwards", 
                   asp = 1)

colo = rep("x", length(Nt_Tree$tip.label))
for (i in seq(1:length(Nt_Tree$tip.label))){
  if (grepl("Lv_TR", Nt_Tree$tip.label[i])){
    colo[i] = viridis(4, end = 0.9)[4]
  } else if (grepl("Lv", Nt_Tree$tip.label[i])) {
    colo[i] = viridis(4, end = 0.9)[3]
  } else if (grepl("Lb", Nt_Tree$tip.label[i])) {
    colo[i] = viridis(4, end = 0.9)[1]
  } else if (grepl("AL", Nt_Tree$tip.label[i])) {
    colo[i] = viridis(4, end = 0.9)[2]
  } else {
    colo[i] = NA_character_
  }
}
cols = setNames(colo, Nt_Tree$tip.label)
```

## Plot Phylo to Map

Next we plot the object and save it as a PDF, because in the default plot window it looks quite weird.

```{r Save PhyloToMap}
pdf(file="TreetoMap.pdf") 
plot(obj, 
     type = "phylogram", 
     direction = "downwards", 
     colors = cols, 
     fill = T, 
     col = "moccasin", 
     cex.points = c(1.1, 1), 
     lwd = 1.2, 
     bg = "darkblue", 
     xlim = c(-10, 59), 
     ylim = c(30, 71)) 
invisible(dev.off())
```
```{r Plot PhyloToMap, echo=FALSE, warning=FALSE, fig.width=7.08, fig.height=6.3, dpi=300}
plot(obj, 
     type = "phylogram", 
     direction = "downwards", 
     colors = cols, 
     fill = T, 
     col = "moccasin", 
     cex.points = c(1.1, 1), 
     lwd = 1.2, 
     bg = "darkblue", 
     xlim = c(-10, 59), 
     ylim = c(30, 71)) 
```

## Align labels

Thats already very nice, but let's try aligning the tip labels manually. So first we check which label is the longest, and for the shorter labels we paste "路" as many times as required so that in the end all labels have the same number of characters.

```{r Align Labels}
obj_aligned = obj

# get the longest label
MaxLabelLength = max(nchar(obj$tree$tip.label))

# now loop over the labels and append "路" 
for (i in (seq(1:length(obj$tree$tip.label)))){
  if (nchar(obj$tree$tip.label[i]) != MaxLabelLength){
    dash = paste0(rep("路", each = MaxLabelLength-nchar(obj$tree$tip.label[i])), collapse = "")
    obj_aligned$tree$tip.label[i] = paste0(dash, obj$tree$tip.label[i])
  }
}

for (i in (seq(1:length(rownames(obj$coords))))){
  if (nchar(rownames(obj$coords)[i]) != MaxLabelLength){
    dash = paste0(rep("路", each = MaxLabelLength-nchar(rownames(obj$coords)[i])), collapse = "")
    rownames(obj_aligned$coords)[i] = paste0(dash, rownames(obj$coords)[i])
  }
}

#  now assign the colors to the new names
colo_aligned = rep("x", length(obj_aligned$tree$tip.label))
for (i in seq(1:length(obj_aligned$tree$tip.label))){
  if (grepl("Lv_TR", obj_aligned$tree$tip.label[i])){
    colo_aligned[i] = viridis(4, end = 0.9)[4]
  } else if (grepl("Lv", obj_aligned$tree$tip.label[i])) {
    colo_aligned[i] = viridis(4, end = 0.9)[3]
  } else if (grepl("Lb", obj_aligned$tree$tip.label[i])) {
    colo_aligned[i] = viridis(4, end = 0.9)[1]
  } else if (grepl("AL", obj_aligned$tree$tip.label[i])) {
    colo_aligned[i] = viridis(4, end = 0.9)[2]
  } else {
    colo_aligned[i] = NA_character_
  }
}
cols_aligned = setNames(colo_aligned, obj_aligned$tree$tip.label)
```

## Plot with aligned labels

```{r Save with aligned labels, message=FALSE, warning=FALSE, results='hide'}
font_import(prompt = F, pattern = "DejaVuSans")
loadfonts()
#pdf(file="TreetoMap_alignedLabels.pdf", family = "DejaVu Sans Mono")
png(file="TreetoMap_alignedLabels.png", family = "DejaVu Sans Mono", res = 300, width = 18, height = 16, units = "cm")
plot(obj_aligned, 
     type = "phylogram", 
     direction = "downwards", 
     colors = cols_aligned, 
     ftype = "b", 
     cex.points = c(1.2, 1.3), 
     cex.text = 1.2,
     lwd = 3)
par(family = "DejaVu Sans")
legend(43.5, 63, legend = c(expression(italic("bilineata")), "Adriatic",
                          expression(italic("viridis")), 
                          expression(italic("viridis")*" (Turkey)"), 
                          "Outgroups"), 
       pt.bg = c(viridis(4, end = 0.9), "white"),
       col = "black", 
       title = expression(bold(bolditalic("Lacerta")*" clades")),
       bg = "grey85", pch = 21)
invisible(dev.off())
```
```{r Plot with aligned labels, echo=FALSE, warning=FALSE, fig.width=7.08, fig.height=6.3, dpi=300}
plot(obj_aligned, 
     type = "phylogram", 
     direction = "downwards", 
     colors = cols_aligned, 
     ftype = "b", 
     cex.points = c(1.2, 1.3), 
     cex.text = 1.2,
     lwd = 3)
par(family = "DejaVu Sans")
legend(43.5, 63, legend = c(expression(italic("bilineata")), "Adriatic",
                          expression(italic("viridis")), 
                          expression(italic("viridis")*" (Turkey)"), 
                          "Outgroups"), 
       pt.bg = c(viridis(4, end = 0.9), "white"),
       col = "black", 
       title = expression(bold(bolditalic("Lacerta")*" clades")),
       bg = "grey85", pch = 21)
```

# Plot phylogram with `ggtree`
## Load tree

Now we use `ggplot` and `ggtree` to plot just the tree, with bootstrap support given by colored node points.

```{r Plot regular tree, warning=FALSE,fig.width = 8.5, fig.height = 8.5, dpi = 300}
Nt_Tree_phylo = read.tree("../00_Data/Trees/RAxML_Nucleotide_Phylogram.nwk")

# sort tip labels so they are in the same order as in the phylo.to.map object
Nt_Tree_phylo = ape::rotateConstr(Nt_Tree_phylo, rev(obj$tree$tip.label))


colo[is.na(colo)] = "black"

g = ggtree(Nt_Tree_phylo, size = 1.1) + 
  geom_tiplab(align = T, hjust = 1, linetype = "dotted", 
              fontface = "bold", offset = 0.2, 
              geom = "label", color = colo, fill = "white", 
              family = "DejaVu Sans Mono", size = 4.5) + 
  geom_nodepoint(aes(color = as.numeric(label)), size = 4.5) + 
  scale_color_viridis_c(option = "cividis", limits = c(0, 100), 
                        direction = -1, name = "Bootstrap Support", 
                        na.value = NA) +
  geom_cladelabel(node = 48, extend = 1, 
                  label = expression(bold("Outgroups")), align = T, 
                  geom = "label", fill = "white", 
                  offset = 0.3, barsize = 2, 
                  family = "DejaVu Sans") +
  geom_cladelabel(node = 30, extend = 0.25, 
                  label = expression(bold(atop(paste(bolditalic("Lacerta viridis")), paste("(Turkey)")))), 
                  align = T, 
                  geom = "label", fill = viridis(4, end = 0.9)[4], 
                  color = c(viridis(4, end = 0.9)[4], "white"), 
                  offset = 0.3, parse = T, barsize = 2, 
                  family = "DejaVu Sans") + 
  geom_cladelabel(node = 32, extend = 0.25, 
                  label = 'bolditalic("Lacerta viridis")', align = T, 
                  geom = "label", fill = viridis(4, end = 0.9)[3], 
                  color = c(viridis(4, end = 0.9)[3], "white"), 
                  offset = 0.3, parse = T, barsize = 2, 
                  family = "DejaVu Sans") + 
  geom_cladelabel(node = 12, extend = 1.5, 
                  label = expression(bold("Adriatic clade")), align = T, 
                  geom = "label", fill = viridis(4, end = 0.9)[2], 
                  color = c(viridis(4, end = 0.9)[2], "white"), 
                  offset = 0.3, barsize = 2, 
                  family = "DejaVu Sans") + 
  geom_cladelabel(node = 44, extend = 1, 
                  label = 'bolditalic("Lacerta bilineata")', align = T, 
                  geom = "label", fill = viridis(4, end = 0.9)[1], 
                  color = c(viridis(4, end = 0.9)[1], "white"), 
                  offset = 0.3, parse = T, barsize = 2, 
                  family = "DejaVu Sans") + 
  xlim(c(-0.05, 1.25)) +
  scale_y_reverse()  + 
  theme_tree() +
  theme(legend.position = c(0.2, 0.4), 
        legend.text = element_text(size = 16), 
        legend.title = element_text(face = "bold", size = 18, 
                  family = "DejaVu Sans")) +
  geom_treescale(x = 0, y = -7.5, width = 0.1) + 
  annotate("text", x = 0, y = 23.2, 
           label = expression(bold("Note on sample labels\n(excl. Outgroups):")), 
           parse = T, hjust = 0, family = "DejaVu Sans", size = 5) + 
  annotate("label", x = 0, y = 24.6, 
           label = expression(paste(atop("9004" , 1), atop("_Lv" , 2), atop("_TR" , 3))), 
           parse = T, hjust = 0, family = "DejaVu Sans Mono", size = 5) + 
  annotate("text", x = 0, y = 27.5, 
           label = "1: Genbank/Tissuecollection ID (if available)\n2: Clade (Lb bilineata, Lv viridis, AL adriatic)\n3: ISO Country Code", 
           hjust = 0, family = "DejaVu Sans", size = 5)

g

ggsave(plot = g, filename = "Tree.pdf", device = cairo_pdf, 
       width = 8.5, height = 8.5, dpi = 300, units = "in")
ggsave(plot = g, filename = "Tree.png", device = "png", 
       width = 8.5, height = 8.5, dpi = 300)

```
## Combine Tree and Phylo to Map

We can plot the tree and the phylomap next to each other by assigning a function to the r base plot and then use this within the `plot_grid` function.

```{r combine Tree and PhyloToMap, warning=FALSE, fig.width = 16, fig.height = 9, dpi = 600}
PhyloMap = function(){
par(family = "DejaVu Sans Mono")
plot(obj_aligned, 
     type = "phylogram", 
     direction = "downwards", 
     colors = cols_aligned, 
     ftype = "b", 
     cex.points = c(1.2, 1.3), 
     lwd = 3)
par(family = "DejaVu Sans")
legend(43.5, 63, legend = c(expression(italic("bilineata")), "Adriatic",
                          expression(italic("viridis")), 
                          expression(italic("viridis")*" (Turkey)"), 
                          "Outgroups"), 
       pt.bg = c(viridis(4, end = 0.9), "white"),
       col = "black", 
       title = expression(bold("Lacerta lineages")),
       bg = "grey85", pch = 21)
}

combi = plot_grid(g, PhyloMap, labels = "AUTO", 
                  nrow = 1, ncol = 2, 
                  label_size = 18, label_fontfamily = "DejaVu Sans", 
                  vjust = 3)

combi

ggsave(plot = combi, filename = "Tree_PhyloToMap_Combined.pdf", 
       device = cairo_pdf, width = 16, height = 9, 
       dpi = 600)
ggsave(plot = combi, filename = "Tree_PhyloToMap_Combined.png", 
       device = "png", width = 16, height = 9, 
       dpi = 600)
```